python/ql/lib/semmle/python/dataflow/new/internal/DataFlowPrivate.qll

```diff
@@ -299,6 +300,14 @@ module LocalFlow {
       nodeTo.(CfgNode).getNode() = def.getDefiningNode()
     )
     or
+    // Assignment to captured variables
+    // These are not covered by the `AssignmentDefinition`s in the case above,
+    // as they are not necessarily live.
+    nodeFrom.(CfgNode).getNode() = nodeTo.(CfgNode).getNode().(DefinitionNode).getValue() and
+    nodeTo.asExpr() = any(VariableCapture::CapturedVariable c).getAStore() and
+    // Exclude assignments to parameters. These are from default values and not local.
+    not nodeTo instanceof ParameterNode
```

Comment:
What is the purpose of this logic?

Reply:
Whenever we have an assignment to a variable, such as `v = 42`, we want flow from the right-hand side to the left-hand side. Most of that is covered by `AssignmentDefinition`s in the lines 298-300 above, but those only exist if `v` is determined to be live. And if `v` is a captured variable, it will not be considered live. So we add this step explicitly for captured variables. There is one exception to take into account, which is default values for functions:
```
def f(x = []):
  pass
```
here `x` is in the scope of `f`, but `[]` is actually evaluated in the outer scope (leading to possible bugs, we have a query for that). So the step from `[]` to `x` is not a local step.

Reply:
But shouldn't writes to captured variables be taken into account by the shared variable capture library (as stores)?

Reply:
You are right! There was a bug in `VariableWrite`; it pointed to the variable instead of the value being written. This also means that we exclude unneeded writes for free :-)